### ISSUE TITLE
[FEATURE REQ] Azure Function EventHub allow passing of the PartitionKey in the EventData ctor

### ISSUE THREAD

**Issue by fuzzlebuck** [Azure/azure-sdk-for-net#33697]
### Library name

Azure.Messaging.EventHubs

### Please describe the feature.

It would be great if could pass a PartitionKey while creating a new EventData object.

This would allow the use of user defined partition keys when adding message via the EventHub output binding's `IAsyncCollector.AddAsync` method.

This is desirable over the use of `EventHubProducerClient.SendAsync` due to the fact that `IAsyncCollector.AddAsync` will queue the message and batch behind the scenes in parallel, whereas `EventHubProducerClient.SendAsync` needs to be awaited on each send of a single message or a batch.

The two mechanisms behave quite differently from each other and we've seen large performance gains while using the `IAsyncCollector` and `IAsyncCollector.Flush` compared to `EventHubProducerClient.SendAsync`.

```cs
IAsyncCollector collector = from output binding
EventData message = new EventData(data);
await collector.AddAsync(message);      <-- returns immediately and is sent in parallel
...
more stuff
...
await collector.Flush();      <-- waits for remaining messages to send
```
vs
```cs
EventHubProducerClient client = from output binding
EventData message = new EventData(data);
await client.SendAsync(message);      <-- waits for the send operation before returning
```

This is similar to some of the comments in https://github.com/Azure/azure-sdk-for-net/issues/28245 but it's not the same as the issue itself.

We've done some testing using the `EventHubsModelFactory.EventData` testing method and verified the partition key when set via this method appears to be used correctly by `IAsyncCollector` and the events do in fact send and arrive downstream in the correct batches with the set partition.

Clearly we can't use this testing and mocking method for production though. 

Or maybe there's another way to use the `IAsyncCollector`?

Thanks :-)

**Comment by jsquire** [comment]
Hi @fuzzlebuck.   Thank you for reaching out and we regret that you're experiencing difficulties.    The `PartitionKey` property of `EventData` is a broker-owned field that is populated only for events read from Event Hubs.   It is intentionally not possible to set a partition key on an event created otherwise.

The implementation of the async collector appears to be incorrect, relying on what was a bug in the previous generation of the Event Hubs client library.  Because of this, partition keys are unable to be used with the collector, despite it attempting to manage batches correctly to do so.   To support this will require a new `AddAsync` overload that accepts a partition key alongside the event.

I'm not sure why this was previously overlooked, but I agree that it's a bug that should be fixed.   We'll move this into the backlog as a bug and see about scheduling it.

**Comment by jsquire** [comment]
//cc: @JoshLove-msft, @m-redding

**Comment by jsquire** [comment]
This has been addressed via #33858.   The release is tracked in #33919 and has been tentatively scheduled for our March window.   Assuming that nothing requires the release to be rescheduled, the updated package should be available in the second week of March.

**Comment by fuzzlebuck** [comment]
Amazing! Thank you! :-)
